File: android/app/src/main/java/sh/mo/ErrorMessages.kt
package sh.mo

/**
 * Error → human message mapping for the UI (one language, consistent:
 * PLAN §4 T6 item "Errors must land in the UI as text"). Pure Kotlin: the
 * core error arrives at the FFI boundary as a `Throwable` whose message is
 * the Rust `#[error(...)]` string; we classify on the UniFFI exception
 * CLASS first (T11: several variants carry an EMPTY message — the empty-jar
 * dead-end — so the message alone cannot be trusted), then on the stable
 * Rust message substrings.
 */
object ErrorMessages {

    /**
     * Map any Throwable from the core bindings to a one-line UI message.
     * The UniFFI class name is matched on `t::class` when it is a generated
     * `CoreException` (no UniFFI import here — the pure suite links this
     * file), so the JVM test drives it with the real class names.
     */
    fun of(t: Throwable): String = when {
        // T11: class-first. `SessionExpired` / `InvalidCredentials` carry an
        // empty message — the string checks below would never see them.
        t.classNameEndsWith("SessionExpired") -> "Session expired — enter the password again."
        t.classNameEndsWith("InvalidCredentials") -> "Wrong username or password."
        t.classNameEndsWith("UpgradeRejected") -> "The gateway rejected the connection (host/origin guard)."
        t.classNameEndsWith("RateLimited") -> "Too many attempts — wait a moment and retry."
        t.classNameEndsWith("NotConnected") -> "Not connected to the gateway yet."

        t.message?.contains("invalid QR payload") == true -> "That pairing payload is not valid."
        t.message?.contains("invalid endpoint url") == true -> "That gateway URL is not valid."
        t.message?.contains("invalid credentials") == true -> "Wrong username or password."
        t.message?.contains("session expired") == true -> "Session expired — enter the password again."
        t.message?.contains("rate limited") == true -> "Too many attempts — wait a moment and retry."
        t.message?.contains("upgrade rejected") == true -> "The gateway rejected the connection (host/origin guard)."
        t.message?.contains("not connected") == true -> "Not connected to the gateway yet."
        t.message?.contains("operation timed out") == true -> "The gateway did not answer in time."
        t.message?.contains("network failure") == true -> "Network error — check the connection."
        t.message?.contains("unknown auth provider") == true -> "The gateway has no password auth enabled."
        else -> "Error: ${t.message ?: t::class.simpleName ?: "unknown"}"
    }

    /**
     * True when the Throwable's simple class name ends with [suffix] — the
     * generated UniFFI `CoreException` nested classes read `SessionExpired`
     * at runtime. `javaClass.simpleName` on a nested Kotlin class is the
     * bare name (`SessionExpired`), so a suffix match survives both the
     * JVM and Android class layouts.
     */
    private fun Throwable.classNameEndsWith(suffix: String): Boolean =
        javaClass.simpleName.endsWith(suffix)

    /**
     * T11: is this failure auth-shaped? True for the UniFFI classes whose
     * fix is re-login (`SessionExpired` — the empty cookie jar, a 401 on a
     * gated route, `InvalidCredentials`). A transport failure is NOT
     * auth-shaped even when its message is empty. Pure Kotlin, so the JVM
     * suite drives the same classifier the repository runs.
     */
    fun isAuthShape(t: Throwable): Boolean =
        t.classNameEndsWith("SessionExpired") ||
            t.classNameEndsWith("InvalidCredentials") ||
            t.classNameEndsWith("UpgradeRejected")
}

File: ios/Sources/HermoGateway/ErrorMessages.swift
import Foundation
import HermesCore

public enum ErrorMessages {

    /// The Rust `#[error(...)]` Display text, which the Kotlin reads straight off
    /// `Throwable.message` and folds into the Offline screen's reason. The generated
    /// `CoreError.errorDescription` is `String(reflecting:)`, so it yields
    /// `HermesCore.CoreError.SessionExpired` and cannot stand in for it.
    public static func rustMessage(_ error: Error) -> String {
        guard let coreError = error as? CoreError else {
            return (error as? LocalizedError)?.errorDescription ?? "?"
        }
        switch coreError {
        case .InvalidQr: return "invalid QR payload"
        case .NotConnected: return "not connected"
        case .Timeout: return "operation timed out"
        case .Rpc(let code, let detail): return "rpc error \(code): \(detail)"
        case .Io(let detail): return "io error: \(detail)"
        case .Network(let detail): return "network failure: \(detail)"
        case .InvalidCredentials: return "invalid credentials"
        case .SessionExpired: return "session expired — re-login required"
        case .UpgradeRejected: return "upgrade rejected by gateway"
        case .InvalidEndpoint(let detail): return "invalid endpoint url: \(detail)"
        case .RateLimited: return "rate limited"
        case .UnknownProvider: return "unknown auth provider"
        case .UnexpectedStatus: return "unexpected submit status"
        case .Http(let status): return "http status \(status)"
        }
    }

    public static func of(_ error: Error) -> String {
        if let coreError = error as? CoreError {
            switch coreError {
            case .SessionExpired:
                return "Session expired — enter the password again."
            case .InvalidCredentials:
                return "Wrong username or password."
            case .UpgradeRejected:
                return "The gateway rejected the connection (host/origin guard)."
            case .RateLimited:
                return "Too many attempts — wait a moment and retry."
            case .NotConnected:
                return "Not connected to the gateway yet."
            case .InvalidQr:
                return "That pairing payload is not valid."
            case .InvalidEndpoint:
                return "That gateway URL is not valid."
            case .Timeout:
                return "The gateway did not answer in time."
            case .Network:
                return "Network error — check the connection."
            case .UnknownProvider:
                return "The gateway has no password auth enabled."
            case .Rpc(let code, let detail):
                return "Error: rpc error \(code): \(detail)"
            case .Io(let detail):
                return "Error: io error: \(detail)"
            case .UnexpectedStatus:
                return "Error: unexpected submit status"
            case .Http(let status):
                return "Error: http status \(status)"
            }
        }

        let message = (error as? LocalizedError)?.errorDescription ?? String(describing: error)
        if message.contains("invalid QR payload") {
            return "That pairing payload is not valid."
        }
        if message.contains("invalid endpoint url") {
            return "That gateway URL is not valid."
        }
        if message.contains("invalid credentials") {
            return "Wrong username or password."
        }
        if message.contains("session expired") {
            return "Session expired — enter the password again."
        }
        if message.contains("rate limited") {
            return "Too many attempts — wait a moment and retry."
        }
        if message.contains("upgrade rejected") {
            return "The gateway rejected the connection (host/origin guard)."
        }
        if message.contains("not connected") {
            return "Not connected to the gateway yet."
        }
        if message.contains("operation timed out") {
            return "The gateway did not answer in time."
        }
        if message.contains("network failure") {
            return "Network error — check the connection."
        }
        if message.contains("unknown auth provider") {
            return "The gateway has no password auth enabled."
        }
        return "Error: \(message)"
    }

    public static func isAuthShape(_ error: Error) -> Bool {
        guard let coreError = error as? CoreError else { return false }
        switch coreError {
        case .SessionExpired, .InvalidCredentials, .UpgradeRejected:
            return true
        default:
            return false
        }
    }
}
